hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java

```diff
@@ -46,36 +48,130 @@
 
+    static class MCEventStore {
+
+        /**
+         * Stores data about a specific MC's next and previous event poll calls
+         */
+        static class LastPollRecord {
+            /**
+             * Denotes when did the MC poll the events last time
+             */
+            final long lastAccessTime;
+            /**
+             * Denotes what sequence should be used when MC polls next time when the events are obtained using
+             * {@link ConcurrentArrayRingbuffer#copyFrom(long)}
+             */
+            final long nextSequence;
+
+            LastPollRecord(long lastAccessTime, long nextSequence) {
+                this.lastAccessTime = lastAccessTime;
+                this.nextSequence = nextSequence;
+            }
+        }
+
+        static final long MC_EVENTS_WINDOW_MILLIS = TimeUnit.SECONDS.toMillis(30);
+        private final LongSupplier clock;
+        private volatile long lastMCEventsPollMillis;
+        private volatile long lastCleanupTimestamp;
+        private final ConcurrentMap<Address, LastPollRecord> lastPollRecordPerMCAddress = new ConcurrentHashMap<>();
+        private final ConcurrentArrayRingbuffer<Event> mcEvents;
+
+        MCEventStore(LongSupplier clock, ConcurrentArrayRingbuffer<Event> mcEvents) {
+            this.clock = clock;
+            this.lastMCEventsPollMillis = clock.getAsLong();
+            this.lastCleanupTimestamp = this.lastMCEventsPollMillis;
+            this.mcEvents = mcEvents;
+        }
+
+        void log(Event event) {
+            if (clock.getAsLong() - lastMCEventsPollMillis > MC_EVENTS_WINDOW_MILLIS) {
+                // ignore event and clear the queue if the last poll happened a while ago
+                onMCEventWindowExceeded();
+            } else {
+                mcEvents.add(event);
+                cleanUpLastAccessRecords();
+            }
+        }
+
+        private static boolean isOutOfTimeWindow(long nowInMillis, long subjectTimestampInMillis) {
+            return nowInMillis - subjectTimestampInMillis > MC_EVENTS_WINDOW_MILLIS;
+        }
+
+        /**
+         * Removing {@link #lastPollRecordPerMCAddress} entries older than {@link #MC_EVENTS_WINDOW_MILLIS} according to their
+         * {@code lastAccessTime}.
+         * <p>
+         * No-op if less than {@link #MC_EVENTS_WINDOW_MILLIS} millis passed since the last cleanup. This avoids too frequent
+         * unnecessary iterations on the {@link #lastPollRecordPerMCAddress} entries.
+         */
+        private void cleanUpLastAccessRecords() {
+            long now = clock.getAsLong();
+            if (!isOutOfTimeWindow(now, lastCleanupTimestamp)) {
+                return;
+            }
+            lastPollRecordPerMCAddress.entrySet().removeIf(entry -> isOutOfTimeWindow(now, entry.getValue().lastAccessTime));
+            lastCleanupTimestamp = now;
+        }
+
+        void onMCEventWindowExceeded() {
+            mcEvents.clear();
```

Comment:
The main benefit of using a ringbuffer is its bounded size and no litter generated by writes. We can always write to the buffer now. With this approach, when a MC instance connects for the first time (or after a disconnect) it receives what was accumulated in the ringbuffer. WAN events objects are relatively compact, so the memory footprint should be small.

@blazember do you see any problems with this approach?